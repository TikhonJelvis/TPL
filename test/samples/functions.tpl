f := \ x -> x
print (f 42)
print (f f f f f f f f f f f 42)

f x := x
print (f 42)
print (f f f f f f f f f f f 42)

f a b c := a + b * c
print (f 1 2 3)

a `f` b := a + b
print (f 1 2)
print (1 `f` 2)

print (typeof f)
print (typeof (f 1))

y := 0
f x := y <- x
print y
f 10
print y
f 42
print y

z := 0
f x := (
    z <- z + 1
    z + x
)
print z
print (f 1)
print (f 1)
print (f 2)

f $x := 42
f (print "fail")

f x $y := x
print (f 42 (print "fail"))

f [x, xs...] := x
print (f [])
print (f [1])
print (f [1,2,3])

f [x, xs...] := xs
print (f [])
print (f [1])
print (f [1,2,3])

f [[a, b], [as, bs]...] := a
print (f [[1, 2], [3, 4]])

f [[a, b], [as, bs]...] := as
print (f [[1,2], [3,4], [5,6], [7,8], [9,10], [11,12]])
