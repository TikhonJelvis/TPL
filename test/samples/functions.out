42
42
42
42
7
3
3
"function"
"function"
0
10
42
0
2
3
5
42
null
1
1
null
null
[2, 3]
1
[3, 5, 7, 9, 11]
null
